File: tests/test_overload_lib.py
"""Library with polymorphic functions for testing overload generation."""


def format_value(value):
    if isinstance(value, str):
        return {"formatted": value.upper(), "kind": "string"}
    if isinstance(value, (int, float)):
        return {"formatted": f"{value:.2f}", "kind": "number"}
    if isinstance(value, bool):
        return {"formatted": "yes" if value else "no", "kind": "boolean"}
    return {"formatted": str(value), "kind": "unknown"}


def convert(input_val, target):
    if target == "string":
        return str(input_val)
    if target == "number":
        return float(input_val)
    return input_val


class Parser:
    def parse(self, input_val):
        if isinstance(input_val, str):
            tokens = input_val.split(" ")
            return {"tokens": tokens, "count": len(tokens)}
        if isinstance(input_val, list):
            return {"tokens": input_val, "count": len(input_val)}
        return {"tokens": [], "count": 0}

File: test-overload-lib.js
/**
 * Library with polymorphic functions that accept different argument types.
 * Used for testing overload generation.
 */

function format(value) {
  if (typeof value === 'string') {
    return { formatted: value.toUpperCase(), kind: 'string' };
  }
  if (typeof value === 'number') {
    return { formatted: value.toFixed(2), kind: 'number' };
  }
  if (typeof value === 'boolean') {
    return { formatted: value ? 'yes' : 'no', kind: 'boolean' };
  }
  return { formatted: String(value), kind: 'unknown' };
}

function convert(input, target) {
  if (target === 'string') {
    return String(input);
  }
  if (target === 'number') {
    return Number(input);
  }
  return input;
}

class Parser {
  parse(input) {
    if (typeof input === 'string') {
      return { tokens: input.split(' '), count: input.split(' ').length };
    }
    if (Array.isArray(input)) {
      return { tokens: input, count: input.length };
    }
    return { tokens: [], count: 0 };
  }
}

module.exports = { format, convert, Parser };
